PortalSwift/Sources/PortalSwift/Provider/PortalProvider.swift

```diff
@@ -0,0 +1,702 @@
+   * Private functions
+   *******************************************/
+
+  private func dispatchConnect() {
+    if let chainId = chainId {
+      let hexChainId = String(format: "%02x", chainId)
+      _ = self.emit(event: Events.Connect.rawValue, data: ["chainId": hexChainId])
+    }
+  }
+
+  private func getApproval(_: String, forPayload: PortalProviderRequestWithId, connect: PortalConnect? = nil) async throws -> Bool {
+    if self.autoApprove {
+      return true
+    }
+    if connect == nil && self.events[Events.PortalSigningRequested.rawValue] == nil {
+      throw ProviderSigningError.noBindingForSigningApprovalFound
+    }
+
+    return try await withCheckedThrowingContinuation { [self] (continuation: CheckedContinuation<Bool, Error>) in
+      _ = self.on(event: Events.PortalSigningApproved.rawValue, callback: { approved in
+        if approved is PortalProviderRequestWithId {
+          let approvedPayload = approved as! PortalProviderRequestWithId
+
+          if approvedPayload.id == forPayload.id, !self.processedRequestIds.contains(forPayload.id) {
+            self.processedRequestIds.append(forPayload.id)
+
+            // If the approved event is fired
+            continuation.resume(returning: true)
+          }
+        }
+      }).on(event: Events.PortalSigningRejected.rawValue, callback: { approved in
+        if approved is ETHRequestPayload {
+          let rejectedPayload = approved as! ETHRequestPayload
+
+          if rejectedPayload.id == forPayload.id, !self.processedRequestIds.contains(forPayload.id) {
+            self.processedRequestIds.append(forPayload.id)
+            // If the rejected event is fired
+            continuation.resume(returning: false)
+          }
+        }
+      })
+
+      if connect != nil {
+        _ = connect?.emit(event: Events.PortalSigningRequested.rawValue, data: forPayload)
+      } else {
+        _ = self.emit(event: Events.PortalSigningRequested.rawValue, data: forPayload)
+      }
+    }
+  }
+
+  private func getRpcUrl(_ chainId: String) throws -> String {
+    guard let rpcUrl = rpcConfig[chainId] else {
+      throw PortalProviderError.noRpcUrlFoundForChainId(chainId)
+    }
+
+    return rpcUrl
+  }
+
+  @available(*, deprecated, renamed: "getRpcUrl", message: "Please use the chain agnostic implementation of getRpcUrl()")
+  private func getRpcUrl(_ chainId: Int) throws -> String {
+    return try self.getRpcUrl("eip155:\(chainId)")
+  }
+
+  private func handleRpcRequest(
+    _ chainId: String,
+    withMethod: PortalRequestMethod,
+    andParams: [AnyEncodable]?,
+    forId: String
+  ) async throws -> PortalProviderResult {
+    let rpcUrl = try getRpcUrl(chainId)
+
+    if let url = URL(string: rpcUrl) {
+      let payload = PortalProviderRpcRequest(
+        id: 0,
+        jsonrpc: "2.0",
+        method: withMethod,
+        params: andParams
+      )
+      let data = try await requests.post(url, withBearerToken: nil, andPayload: payload)
```

Comment:
why is `withBearerToken` nil here?

Reply:
Because these are RPC requests, so the token is a part of the URL.